Improve the dev guide handling

shubham00jain opened the pull request:
Fixes #4536.

I haven't changed index.rst using jinja like @Cadair suggested. Please first review if I'm going in the right direction.

nabobalis commented on docs/conf.py:
```diff
@@ -218,7 +224,12 @@
 # -- Stability Page ------------------------------------------------------------
 
-with open('./dev_guide/sunpy_stability.yaml', 'r') as estability:
+if is_development == True:
+    url = './dev_guide/sunpy_stability.yaml'
+else:
+    url = './dev_guide_release/sunpy_stability.yaml'
```
I am not sure this is what we are after. 

The goal is prevent the removal of the dev_guide when we do a release. 

The stability page should be moved wholesale.

shubham00jain replied:
The goal was to automatically remove all the files inside dev_guide (except index.rst)  whenever a stable version is released. So I made 2 folders one for stability release and one for dev release and this url tells the location of the `stability.yaml` file. Is this approach wrong?

nabobalis replied:
Does this work? 

I think we want to move the sunpy_stability page regardless of if its a release or not.

nabobalis commented:
I think two things need to be done:

1. Move the stability.rst to docs/code_ref and that should be where it lives, no matter if it is a release or development
2. Remove the extra files, those pngs and the extra stability file.

shubham00jain commented:
Things I've done:
1. The stability files only lives inside code_ref now.
2. Removed extra png images from dev_guide_release. 
3. Updated the index.rst to automatically include whichever version we're on.

Things remaining:
> Activate a redirect on read the docs to automatically redirect people who visit the stable URL to the latest one. (This will have to be done by a project admin one the changes have been merged and backported).

This will have to been done by an admin.

nabobalis commented on docs/dev_guide_release/index.rst:
```diff
@@ -0,0 +1,12 @@
+
+Please go <here> for our up to date developer’s guide.
+
+.. toctree::
+   :maxdepth: 2
+
+   ../code_ref/stability
```
This does not need to be linked by the dev guide.

nabobalis replied:
It does need to be linked by the index file for the code ref.

nabobalis commented on docs/dev_guide/index.rst:
```diff
@@ -26,4 +26,4 @@ This goes over the basics and has links to useful tutorials on git.
    config
    funding
-   stability
+   ../code_ref/stability
```
```suggestion
```
This should linked in the index file for the code_ref.

nabobalis commented on docs/dev_guide_release/index.rst:
```diff
@@ -0,0 +1,12 @@
+Developer's Guide
+*****************
+
+Please go <here> for our up to date developer’s guide.
```
How is here dealt with?

shubham00jain replied:
<here> should be linked with the updated dev_guide version (most probably by an admin).

nabobalis replied:
I am not sure I follow.

Cadair replied:
This should be the same link as it is now.

shubham00jain replied:
Got it. I'll link that.

nabobalis commented:
There are still extra png images.

nabobalis commented on docs/conf.py:
```diff
@@ -115,8 +115,14 @@
+
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
 
+if is_development == True:
```
```suggestion
if is_development:
```

nabobalis commented on docs/dev_guide_release/index.rst:
```diff
@@ -0,0 +1,12 @@
+
+Please go <here> for our up to date developer’s guide.
+
+.. toctree::
+   :maxdepth: 2
+
+   ../code_ref/stability
```
```suggestion
```

nabobalis commented on docs/index.rst:
```diff
@@ -18,3 +18,13 @@ Documentation :any:`index <genindex>` and :any:`Module <modindex>` index.
   whatsnew/index
   about
+
+.. toctree::
+  :maxdepth: 2
+
+  guide/index
+  code_ref/index
+  generated/gallery/index
+  dev_guide_release/index
+  whatsnew/index
+  about
```
Why was this added?

shubham00jain replied:
I added this because I thought sphinx will need everything again for the toctree of release version. I just built it and checked, it's unnccesory and only `dev_guide_release/index` needs to be added after `dev_guide/index`. Fixed.

nabobalis commented:
You now seem to have removed the original images, we kind of want those.

nabobalis commented on docs/conf.py:
```diff
@@ -247,4 +253,4 @@ def setup(app):
 
     # The theme conf provides a fix for circle ci redirections
-    fix_circleci(app)
+    fix_circleci(app)
```
This change needs to be reverted.

nabobalis commented:
I think with a bit more tidy, it will be ready.

shubham00jain commented:
The only thing remaining is linking <here> in dev_guide_release.

nabobalis commented on docs/index.rst:
```diff
@@ -16,5 +16,6 @@ Documentation :any:`index <genindex>` and :any:`Module <modindex>` index.
   code_ref/index
   generated/gallery/index
   dev_guide/index
+  dev_guide_release/index
```
How does this work? If one is excluded at build time, I thought sphinx would error due to a missing reference.

shubham00jain replied:
Sphinx just ignores it. I've built it and checked.

Cadair replied:
This isn't what I had in mind, you should be able to remove this once you get the conditional stuff working in the index.rst. That being said if sphinx wont error for this then it will probably make your template logic simpler.

nabobalis replied:
Sphinx does raise a warning, which we try to avoid.

shubham00jain replied:
So, what should be the alternative? Should the whole thing be written using jinga?

nabobalis replied:
Good question, it might be worth trying.

nabobalis replied:
I think only the index file would need to have jinja templating.

Cadair replied:
yeah, you can template the index file so if it's a dev version the toctree is included and if it isn't then just the link is shown

shubham00jain commented:
@nabobalis @Cadair  I tried using jinja2 template in index.rst and I was not able make it work correctly. Can you tell what is needed to be done?

nabobalis commented:
> @nabobalis @Cadair I tried using jinja2 template in index.rst and I was not able make it work correctly. Can you tell what is needed to be done?

What did you in the index file?

shubham00jain commented:
@nabobalis 
```
{% raw %}
.. raw:: html
{% endraw %}

{% from .conf import is_development %}

.. toctree::
  :maxdepth: 2

  guide/index
  code_ref/index
  generated/gallery/index
  {if is_development:}
    dev_guide/index
  {else:}
    dev_guide_release/index
  whatsnew/index
  about
```
I changed the toctree to this. I don't think this is correct. Can I use toctree in jinja2 template or do I have to make the whole ul->li->a loop for every folder?

nabobalis commented:
As long as it gets templated, I don't see why you would need to do anything else.

shubham00jain commented:
The if and else statements aren't getting templated and are shown as plain text.

nabobalis commented:
Will probably need to figure out how to get jina tempting to pick that up page, however we did it for the stability page.

shubham00jain commented:
@Cadair @nabobalis I have tried many approaches but wasn't able to get the desired jinja templating on `index.rst` file. Can you guys please help?

nabobalis commented:
> @Cadair @nabobalis I have tried many approaches but wasn't able to get the desired jinja templating on `index.rst` file. Can you guys please help?

Where you able to get the templating to work at all?

shubham00jain commented:
@nabobalis No. If I copy the exact content as in 'stability.rst' file into `index.rst` file, I get the templated file but I make any changes into it the template goes away.

nabobalis commented:
Can you commit what you have and I can try to have a look.

shubham00jain commented:
@nabobalis I tried to use the `.. only::` directive and `ifconfig` extension but they don't seem to work as well.

shubham00jain commented:
@nabobalis Hi. Did you get a chance to look at my commit?

nabobalis commented:
Sorry no, I will have a go this evening after work.

Cadair commented:
I think the templating isn't working because of this: https://github.com/sunpy/sunpy/blob/master/docs/conf.py#L237

shubham00jain commented:
@nabobalis The templating is working now! 😃  Please tell if I missed anything.

nabobalis commented on docs/conf.py:
```diff
@@ -234,7 +241,7 @@ def rstjinja(app, docname, source):
         return
     src = source[0]
-    if "Current status" in src[:20]:
+    if "Current status" or "SunPy Documentation" in src[:20]:
```
"Sunpy documentation" could hit other pages?

nabobalis commented on docs/dev_guide_release/index.rst:
```diff
@@ -0,0 +1,7 @@
+Developer's Guide
+*****************
+
+Please go `here <https://docs.sunpy.org/en/latest/dev_guide/index.html>`_ for our up to date developer’s guide.
```
```suggestion
Please go `here <https://docs.sunpy.org/en/latest/dev_guide/index.html>`__ for our up to date developer’s guide.
```

nabobalis commented:
Yeah the changes look good.

shubham00jain commented:
@nabobalis I don't think any other file which has "Sunpy Documentation" is templated using jinja, so we should be fine.

nabobalis commented:
You will need to fix the issues the failing python_codestyle is finding.

shubham00jain commented:
@nabobalis Sorry I don't think I follow. What issues?

nabobalis commented:
> @nabobalis Sorry I don't think I follow. What issues?

The "sunpy.sunpy (python_codestyle [linux]) Failing after 1m — python_codestyle [linux] failed" should pass and you will need to solve the problem it found. 

```bash
All changes made by hooks:
diff --git a/docs/conf.py b/docs/conf.py
index bcf1eac8c..febb67db2 100644
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -254,4 +254,4 @@ def setup(app):
     app.connect("source-read", rstjinja)
 
     # The theme conf provides a fix for circle ci redirections
-    fix_circleci(app)
\ No newline at end of file
+    fix_circleci(app)

```

shubham00jain commented:
@nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ?

nabobalis commented:
> @nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ?

Yes.

nabobalis commented:
> @nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ? 
> 

If you run that pre commit command, it should fix the file for you. I am surprised it didn't do anything.

Cadair commented:
I think this looks fine as it is, although I have one thought that might make it a little bit clearer. Instead of templating the toctree in `docs/index.rst`, why don't we template `docs/dev_guide/index.rst`? We can have a block like:

```
{if is_development}
current content of file
{else}
link to dev version
{endif}
```

shubham00jain commented:
@Cadair We can do this but then we won't need to use `exclude_patterns`. Just a simple template in `dev_guide/index.rst` would suffice. So should I make that change?

@nabobalis Whenever I run `pre-commit run --all-files` I get an error that my python/pip version is not update (which I have already done). I have tried to use virtualenv as well. Can you suggest some fix?

nabobalis commented:
> @Cadair We can do this but then we won't need to use `exclude_patterns`. Just a simple template in `dev_guide/index.rst` would suffice. So should I make that change?

Sure.

> @nabobalis Whenever I run `pre-commit run --all-files` I get an error that my python/pip version is not update (which I have already done). I have tried to use virtualenv as well. Can you suggest some fix?

I need more information to debug this. Can I see an log?

shubham00jain commented:
@nabobalis 
This is the log:
 https://pastebin.pl/view/22960518

nabobalis commented:
> @nabobalis
> This is the log:
> [pastebin.pl/view/22960518](https://pastebin.pl/view/22960518)

`ERROR: isort-5.5.2 requires Python '>=3.6,<4.0' but the running Python is 3.5.2 you need to use a more up to date version of pytohn.

shubham00jain commented:
running python3 --version command gives me python 3.8.5

nabobalis commented:
> running python3 --version command gives me python 3.8.5

Somewhere along the way, pre-commit is not using the correct version of python and that's probably a virtual environment issue. You want to run the tox build directly.

Cadair commented:
> @Cadair We can do this but then we won't need to use exclude_patterns. Just a simple template in dev_guide/index.rst would suffice. So should I make that change?

I think we will still need to do this, or else there will be a bunch of rst files which aren't included in a toctree which will cause sphinx to throw warnings.

shubham00jain commented:
@Cadair Can you tell me the exact structure of the docs folder which you want? I tried to store all the contents in a new `dev/mycontents` and then excluding that but the whole dev_guide folder didn't show. I can try doing that again maybe I missed something.

@nabobalis how to run that ?(This is my first 'good' opensource PR, I'm quite new to this.)

Cadair commented:
> @Cadair Can you tell me the exact structure of the docs folder which you want? I tried to store all the contents in a new `dev/mycontents` and then excluding that but the whole dev_guide folder didn't show. I can try doing that again maybe I missed something.

I guess I am trying to avoid having a new top level folder in the docs tree. So put all the real content in `dev_guide/content/` then rewrite the toctree in index.rst to reference things in that subfolder, then exclude the subfolder in conf.py

nabobalis commented on docs/conf.py:
```diff
@@ -115,8 +115,12 @@
+
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
 
+if is_development == False:
```
```suggestion
if not is_development:
```

Cadair commented:
Something seems to have broken the sphinx build: https://dev.azure.com/sunpy/sunpy/_build/results?buildId=10527&view=logs&j=196c4937-6990-58b5-7d18-234ac2f0888b&t=e790bdf0-5cca-5361-ce0e-7cb2a4cbf89d&l=205

> sphinx.errors.ExtensionError: Handler <function rstjinja at 0x7fe919be1ca0> for event 'source-read' threw an exception (exception: unexpected ':')

shubham00jain commented:
@Cadair I just cloned the code from sunpy repo and made all the changes again, this time very carefully without adding/removing any unnecessary white-spaces but as soon as I make these changes and try to build to them, I get error in `generated/gallery/computer_vision_techniques/off_limb_enhance ` which I didn't even touch. Can you suggest what might be wrong?

nabobalis commented:
> @Cadair I just cloned the code from sunpy repo and made all the changes again, this time very carefully without adding/removing any unnecessary white-spaces but as soon as I make these changes and try to build to them, I get error in `generated/gallery/computer_vision_techniques/off_limb_enhance ` which I didn't even touch. Can you suggest what might be wrong?

1. What was the error?
2. You can ignore the gallery and I would suggest not building it when you are testing out the documentation.
   `tox -e build_docs -- -D plot_gallery=False`

shubham00jain commented:
@nabobalis 

I get this ```Extension error:
Handler <function rstjinja at 0x7f01c2a78b80> for event 'source-read' threw an exception (exception: unexpected ':')
make: *** [Makefile:19: html] Error 2``` when I run `make html`. I don't think a **:** is the problem because the same logic runs when I template `docs/index.rst` instead of `docs/dev_guide/index.rst`.

I tried using ```tox -e build_docs -- -D plot_gallery=False``` but I get `ModuleNotFoundError: No module named 'virtualenv.seed.via_app_data'`. This is the same error I get when I tried to use `tox -e codestyle`. For this, I've tried re-installing virtualenv ,python and pip.

And finally I use VSCode with Microsoft's Default Python Linter/Debugger extension installed.

So my question is, is this happening because of indexing a different file?

nabobalis commented:
> I get this `Extension error: Handler <function rstjinja at 0x7f01c2a78b80> for event 'source-read' threw an exception (exception: unexpected ':') make: *** [Makefile:19: html] Error 2` when I run `make html`. I don't think a **:** is the problem because the same logic runs when I template `docs/index.rst` instead of `docs/dev_guide/index.rst`.

How about removing text in the first if block and seeing it compiles? 

> I tried using `tox -e build_docs -- -D plot_gallery=False` but I get `ModuleNotFoundError: No module named 'virtualenv.seed.via_app_data'`. This is the same error I get when I tried to use `tox -e codestyle`. For this, I've tried re-installing virtualenv ,python and pip.

I am not sure how to help here but something is differently not setup correctly with the virtual environment. I just can't quite pin it. 

> So my question is, is this happening because of indexing a different file?

This looks to be the case.

shubham00jain commented:
@Cadair 
I was looking at the error and I have found some problems with putting everything inside `dev_guide/contents` eg. in `config.rst` which was previously inside `dev_guide/` has a **relative link** to another file ` :doc:../guide/customization ` which won't work because we have changed its own path. and sphinx starts giving warning here. There might be some more files like these which I couldn't find which are probably causing the error I'm facing.

So, I would like to suggest my previous approach again i.e using 2 folders. In that way, we won't have to check everything inside every rst file in dev_guide.

nabobalis commented:
Thanks for working on this @shubham00jain. It was a bit more complex than we expected when we opened the original issue.

nabobalis commented:
Still see
```
/home/vsts/work/1/s/docs/dev_guide/index.rst:9: WARNING: duplicate label newcomers, other instance in /home/vsts/work/1/s/docs/dev_guide/contents/newcomers.rst
/home/vsts/work/1/s/docs/dev_guide/index.rst:9: WARNING: duplicate label remote_data, other instance in /home/vsts/work/1/s/docs/dev_guide/contents/remote_data.rst
```

nabobalis commented on docs/dev_guide/index.rst:
```diff
@@ -1,9 +1,15 @@
 .. _dev_guide:
+{% if is_development %}
```
```suggestion
{% if not is_development %}
```
Maybe?

Cadair replied:
yep

nabobalis commented:
Example failed:
```
Extension error:
Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/vsts/work/1/s/examples/plotting/magnetogram_active_regions.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/vsts/work/1/s/examples/plotting/magnetogram_active_regions.py", line 46, in <module>
    srs_table = srs.read_srs(srs_downloaded_files[0])
  File "/home/vsts/work/1/s/.tox/build_docs/lib/python3.8/site-packages/sunpy/io/special/srs.py", line 32, in read_srs
    with open(filepath) as srs:
FileNotFoundError: [Errno 2] No such file or directory: '/home/vsts/work/1/s/.tox/build_docs/tmp/sunpy/data/20110607SRS.txt'
```
Unrelated to the PR.

nabobalis commented:
Thanks for the PR @shubham00jain
